### PR TITLE
Feat: [app/dns] Support per-client configuration

### PR DESCRIPTION
**Pull request by Vigilans**
This PR defines a new way to interpret DNS module and all its name servers: 

Each nameserver (the client) could configure their options (e.g. tag, clientIP, queryStrategy) **independently**, while options specified in the DNS module are used as **global default**.

That is, a client for now could configure following options individually:

**Inbound Tag**
A special nameserver could name its inbound with a unique tag, thus made possible for domain-based DNS routing. 
E.g. a server resolving `geosite:bilibili` gets named `bilibili`, and in routing it could be routed either directly or via HK proxy, to switch between different regions.

**Client IP**
Nameservers of same address could use different `clientIP` for different domains.

**Fallback Strategy**
A special nameserver serving `geosite:bilibili` may not expect to be used by other domains. Thus could be disabled from fallback individually.
On the other hand, when a domain matching `geosite:bilibili`, we also do not want the fallback servers to resolve it anymore, such could be set to `disabledIfAnyMatch`.

**Cache Strategy**
Each nameserver could disable cache individually.

**Query Strategy**
We could force certain domains to use only IPv4 or IPv6 address, if either route is preferred or with better quality.
_A use case in transparent proxy:_ 
In transparent proxy, we could specify all IPv6 packets go direct, but route all global IPv4 addresses to proxy (since when proxied, it normally doesn't matter whether dest is v4 or v6). Then, with `queryStrategy`, we could enforce all **1)** global geosite domains, and **2)** domains hit fallback server expecting non-CN IP to return only IPv4 IP, therefore gets routed to v2ray naturally. 
This sorts of works like fakedns, but all resolved IPs are real, and v2ray does not take control of all traffic, only expected ones. 

In this PR, following options are deprecated (but keep compatible with new options):
* `disableFallback`, `disableFallbackIfMatch`, `skipFallback`: use `fallbackStrategy`.
* `disableCache`: use `cacheStrategy`.

This PR also refactors the DNS module code, majorly to improve the readability and renders it more maintainable.

This PR does not introduce any change to existing features API. They'll be discussed and separated to future PRs.

### CODE REVIEW

**Review by AkinoKaede — APPROVED** [review]

File: app/dns/config.proto [hunk 1]
```diff
@@ -121,13 +148,23 @@ message SimplifiedConfig {
...
   QueryStrategy query_strategy = 9;
 
-  bool disableFallback = 10;
+  // Default cache strategy for each name server.
+  CacheStrategy cache_strategy = 12;
 
-  bool disableFallbackIfMatch = 11;
+  // Default fallback strategy for each name server.
+  FallbackStrategy fallback_strategy = 13;
 }
```
[the remark this answers is not in the document]
Comment (AkinoKaede) [reply app/dns/config.proto:168]: I think marking deprecated is better.

**Comment by Vigilans** [comment]
# Future works

## Routing Strategy

This strategy aims to unify the three scenarios in routing dns inbound requests: 
* `Default`: sends to router normally.
* `SkipDNS`: do not use internal DNS. Corresponds to `session.Content.SkipDNSResolve`.
* `Skip`: do not use router. Corresponds to schemes like `https+local://`.

## FakeDNS

Work on FakeDNS aims to remove `FakeEnable` member in `dns.IPOption`, `ipOption` and `ctx` member in `DNS`, and provide a new user-friendly way to configure FakeDNS.

Currently I may propose two ways and would like to know any of them would be a good idea:

### Configure `fakedns` server objects independently from `servers`

Since fakedns is only enabled when using dns outbound, putting the fakedns server object along with other normal servers made the rules in `servers` hard to interpret when fakedns not enabled.

We may configure the fakedns server objects in a field sibling to `servers` field, and merge them to a full servers list when fakedns is enabled, at a certain priority:
```json5
"dns": {
    "tag": "dns",
    "servers": [
      {
        "address": "8.8.8.8",
        "domains": ["geosite:geolocation-!cn"]
      }
    ],
    "fakedns": {...}
}
```


### Configure `fakedns` as an independent configuration in each nameserver object

As per [FakeDNS | V2Fly.org](https://www.v2fly.org/config/fakedns.html) and https://github.com/v2fly/v2ray-core/discussions/789, integrating fakedns into `servers` is intended for rule-based dns resolution.  In such case, it is the nameserver object that holds the rule semantic, but not the target DNS server address. 

Therefore, we could integrate `fakedns` as a configurable option in a nameserver object, which will return fake ip instead of using original nameserver when fakedns is enabled (also a global default `fakedns` option):
```json5
"dns": {
    "tag": "dns",
    "servers": [
      {
        "address": "8.8.8.8",
        "domains": ["geosite:geolocation-!cn"],
        "fakedns": true
        // or 
        "fakedns": "xxx.xxx.xxx.xxx/xx"  // to return ips in a specific prefix ?
      }
    ]
}
```

## Geoip supports specifying IPv4 or IPv6 subsets (Backlog)

Currently, the expect IP "All ipv4 address and only CN ipv6 address" are conveyed as

```json5
expectIPs: ["0.0.0.0/0", "geoip:cn"] // ipv4 from geoip:cn just included in "0.0.0.0/0"
```

It would be better if we could express them as `["geoip:any@v4", "geoip:cn@v6"]` 
(Yes, also a new country code "any", where `geoip:any@v4 = [0.0.0.0/0]`, `geoip:any@v6 = [::/0]`)

## Automatically creates a "geosite:builtin" domain list for all domains written in DNS and Outbound configs (Backlog)

Maintaining a list of all domains in v2ray config will help resolve may issues related to infinite loop.

## Forward and dispatch non-A and AAA requests (Backlog)

Currently, dns outbound only forwards A and AAAA requests to dns app, and forward all other requests to its original destination. 

DNS app may also take other type of requests, but only does rules dispatching for them, and forwards the raw dns message to the dispatched destination.

## Reserve original destination in DNS outbound (Backlog)

When dns outbound send the dns request to dns app, the latter does not hold any information about the original destination. Sometimes it may help (and simplify the config writing) if dns app has knowledge of the original destination.

**Comment by kslr** [comment]
cc @xiaokangwang

**Comment by AkinoKaede** [comment]
This is a very constructive change and we should review the code and merge it as soon as possible.

**Review by kslr — APPROVED** [review]

**Comment by AkinoKaede** [comment]
I reviewed it again.

**Pull request merged**